hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -524,10 +528,94 @@ def should_generate_more():
+            #
+            # The heuristic we use is that we attempt to estimate the smallest
+            # extension of this prefix, and limit the size to no more than
+            # an order of magnitude larger than that. If we fail to estimate
+            # the size accurately, we skip over this prefix and try again.
+            #
+            # We need to tune the example size based on the initial prefix,
+            # because any fixed size might be too small, and any size based
+            # on the strategy in general can fall afoul of strategies that
+            # have very different sizes for different prefixes.
+            small_example_cap = max(10, min(self.settings.max_examples // 10, 50))
+
+            if (
+                self.valid_examples <= small_example_cap
+                and self.call_count <= 5 * small_example_cap
+            ):
+                # If we didn't get a valid example then we can't reliably
+                # use the length to predict the size of the buffer, so we
+                # just take a rough guess.
+                max_length = max(len(prefix) * 10, BUFFER_SIZE // 10)
+
+                if consecutive_zero_extend_is_invalid < 5:
+                    minimal_example = self.cached_test_function(
+                        prefix + hbytes(BUFFER_SIZE)
+                    )
+
+                    if minimal_example.status >= Status.VALID:
+                        max_length = (
+                            len(minimal_example.buffer) - len(prefix)
+                        ) * 10 + len(prefix)
+                        consecutive_zero_extend_is_invalid = 0
+                    else:
+                        consecutive_zero_extend_is_invalid += 1
+
+                # We might have hit the cap on number of examples we should
+                # run when calculating the minimal example.
+                if not should_generate_more():
+                    break
+            else:
+                max_length = BUFFER_SIZE
+
+            # We could end up in a situation where even though the prefix was
+            # novel when we generated it, because we've now tried zero extending
+            # it not all possible continuations of it will be novel. In order to
+            # avoid making redundant test calls, we rerun it in simulation mode
+            # first. If this has a predictable result, then we don't bother
+            # running the test function for real here. If however we encounter
+            # some novel behaviour, we try again with the real test function,
+            # starting from the new novel prefix that has discovered.
+            try:
+                trial_data = self.new_conjecture_data(
+                    draw_bytes_with(prefix, parameter), max_length=max_length
+                )
+                self.tree.simulate_test_function(trial_data)
+                continue
+            except PreviouslyUnseenBehaviour:
+                pass
```

Comment:
It seems like a waste to run this in iterations where we didn't actually try any zero-extending, either because we're out of the small-example stage or because we tripped the bad-zero-extend heuristic.

Reply:
Ah, that's true.